*****
After gathering scoreboard data, I can get insights into team compositions.   

There have been difficulties with scraping this information, even with the use of the API. This is because websites like tracker.gg employ Cloudfare protection, which means that they actively work against people scraping their data.     

So, since I frequently get rate limited and i have to take breaks/timeouts before attempting again, I have been able to get about 3,666 matches but not the full 5,400, after 15+hrs of scraping. This is not the ideal scenario but it is enough data to still work with.    

I think that in the future, longer periods of time could be allotted to scraping, but since the possibility of being able to scrape this data was not discovered until recently, it is reasonable to work with what I have gotten so far. (upon inspection of other data, it was revealed the internal match ids used by tracker.gg, which is appended at the end of each api call to the scoreboard, which i did not have a way of knowing until this)
*****

In [1]:
import sqlite3
import pandas as pd

conn = sqlite3.connect('full_matches.db')
cursor = conn.cursor()
cursor.execute("SELECT name FROM sqlite_master WHERE type='table';")
tables = cursor.fetchall()

print(tables)
#conn.close()
#matches.head()

[('full_matches',)]


In [2]:
matches = pd.read_sql_query("SELECT * FROM full_matches", conn)
conn.close()
matches.head()

match_id  \
0  5513730_1765074019_1318548_11001_12   
1  6711722_1764888815_1272257_11001_22   
2  6711647_1764886525_1245250_11001_22   
3   506241_1764884213_1311009_11001_21   
4   506105_1764882596_1310017_11001_21   

                                            raw_json  
0  {"data": {"attributes": {"id": "5513730_176507...  
1  {"data": {"attributes": {"id": "6711722_176488...  
2  {"data": {"attributes": {"id": "6711647_176488...  
3  {"data": {"attributes": {"id": "506241_1764884...  
4  {"data": {"attributes": {"id": "506105_1764882...

*****
It appears the same kind of raw json formatting is present as when i unpacked the file in the 'exploratory_analysis.ipynb" notebook. As so, i am going to copy over that code here to do the same thing. 
*****

In [3]:
import json 
matches['raw_json'] = matches['raw_json'].apply(json.loads)
matches

match_id  \
0     5513730_1765074019_1318548_11001_12   
1     6711722_1764888815_1272257_11001_22   
2     6711647_1764886525_1245250_11001_22   
3      506241_1764884213_1311009_11001_21   
4      506105_1764882596_1310017_11001_21   
...                                   ...   
5413  5517996_1765106140_1311312_11001_11   
5414  5517410_1765104631_1291702_11001_11   
5415  5517426_1765100984_1267184_11001_11   
5416  5513998_1765096434_1291511_11001_12   
5417  5513925_1765094719_1231224_11001_12   

                                               raw_json  
0     {'data': {'attributes': {'id': '5513730_176507...  
1     {'data': {'attributes': {'id': '6711722_176488...  
2     {'data': {'attributes': {'id': '6711647_176488...  
3     {'data': {'attributes': {'id': '506241_1764884...  
4     {'data': {'attributes': {'id': '506105_1764882...  
...                                                 ...  
5413  {'data': {'attributes': {'id': '5517996_176510...  
5414  {'data': {'attributes': {'id': '5517410_176510...  
5415  {'data': {'attributes': {'id': '5517426_176510...  
5416  {'data': {'attributes': {'id': '5513998_176509...  
5417  {'data': {'attributes': {'id': '5513925_176509...  

[5418 rows x 2 columns]

In [4]:
print(json.dumps(matches["raw_json"].iloc[0], indent=4))

{
    "data": {
        "attributes": {
            "id": "5513730_1765074019_1318548_11001_12",
            "mode": "competitive",
            "mapId": 1318
        },
        "metadata": {
            "mapName": "Celestial Husk",
            "mapModeName": "Domination",
            "mapImageUrl": "https://trackercdn.com/cdn/tracker.gg/marvel-rivals/images/maps/hero/1318.jpg?v=1756050380",
            "timestamp": "2025-12-07T02:30:35+00:00",
            "duration": 381,
            "modeName": "Competitive",
            "isRanked": true,
            "winningTeamId": 0,
            "scores": [
                2,
                0
            ],
            "replayId": "10858083708",
            "fullMatchAvailable": true
        },
        "segments": [
            {
                "type": "player",
                "attributes": {
                    "accountId": "574bec15-67b9-412e-b988-3634e5a88915"
                },
                "metadata": {
                    "platformInfo"

In [5]:

rows = []

for i, row in matches.iterrows():
    raw_json = row["raw_json"]

    # Skip any missing data w/ continue
    if raw_json is None or "data" not in raw_json:
        continue
    
    data = raw_json["data"]
    match_id = data["attributes"]["id"]

    match_meta = data.get("metadata", {})
    match_map = match_meta.get("mapName")
    match_mode = match_meta.get("mapModeName")
    match_duration = match_meta.get("duration")
    winning_team = match_meta.get("winningTeamId")

    # need to go through all segments to get just the player stats 
    #that i want so i can match with the player data i already have
    for segment in data.get("segments", []):
        if segment.get("type") != "player": #theres data with non player data that idc about so skip
            continue  #skip it


    #originally i did segment['metadata'] bc cGPT said to do so
    # but added get() instead to avoid key errors 
    # so the loop could continue without crashing even if some data is missing

        meta = segment.get("metadata", {})
        stats = segment.get("stats", {})

        player_id = meta.get("platformInfo", {}).get("platformUserIdentifier")
        team_id = meta.get("teamId")
        result = meta.get("result")   #result is win result
        is_mvp = meta.get("isMvp", False) #assume false if there isnt the field there
        is_svp = meta.get("isSvp", False) #assume false if there isnt the field there

        # hero played
        heroes = meta.get("heroes", [])
        hero_name = heroes[0].get("name") if heroes else None
        hero_id = heroes[0].get("heroId") if heroes else None

        # flatten stats like in the other notebook
        flat_stats = {}
        for stat_key, stat in stats.items():
            flat_stats[f"{stat_key}_value"] = stat.get("value")

        rows.append({
            "match_id": match_id,
            "player_id": player_id,
            "team_id": team_id,
            "won": 1 if team_id == winning_team else 0, #need to distinguish meaning of 1/0 with w/l
            "hero_name": hero_name,
            "hero_id": hero_id,
            "map_name": match_map,
            "match_mode": match_mode,
            "duration_sec": match_duration,
            "is_mvp": is_mvp,
            "is_svp": is_svp,
            **flat_stats
        })

matches_cleaned = pd.DataFrame(rows)


In [6]:
matches_cleaned

match_id      player_id  team_id  won  \
0      5513730_1765074019_1318548_11001_12     Deleted404        0    1   
1      5513730_1765074019_1318548_11001_12          Kaifs        0    1   
2      5513730_1765074019_1318548_11001_12         Finnsi        0    1   
3      5513730_1765074019_1318548_11001_12    JukeBox2274        1    0   
4      5513730_1765074019_1318548_11001_12       VP Sypeh        0    1   
...                                    ...            ...      ...  ...   
64051  5513925_1765094719_1231224_11001_12          iffvy        1    0   
64052  5513925_1765094719_1231224_11001_12      Rajahyrax        1    0   
64053  5513925_1765094719_1231224_11001_12            Пyк        1    0   
64054  5513925_1765094719_1231224_11001_12  FAZE BLAZIKEN        0    1   
64055  5513925_1765094719_1231224_11001_12  сountry mouse        1    0   

             hero_name  hero_id         map_name  match_mode  duration_sec  \
0       Rocket Raccoon     1023   Celestial Husk  Domination           381   
1              Magneto     1037   Celestial Husk  Domination           381   
2                Groot     1027   Celestial Husk  Domination           381   
3               Angela     1056   Celestial Husk  Domination           381   
4      Invisible Woman     1050   Celestial Husk  Domination           381   
...                ...      ...              ...         ...           ...   
64051   Winter Soldier     1041  Yggdrasill Path      Convoy          1059   
64052            Groot     1027  Yggdrasill Path      Convoy          1059   
64053   Rocket Raccoon     1023  Yggdrasill Path      Convoy          1059   
64054          Phoenix     1054  Yggdrasill Path      Convoy          1059   
64055          Phoenix     1054  Yggdrasill Path      Convoy          1059   

       is_mvp  ...  featureHitRate2ChaosHits_value  \
0       False  ...                             NaN   
1       False  ...                             NaN   
2       False  ...                             NaN   
3       False  ...                             NaN   
4       False  ...                             NaN   
...       ...  ...                             ...   
64051   False  ...                             NaN   
64052   False  ...                             NaN   
64053   False  ...                             NaN   
64054   False  ...                             NaN   
64055   False  ...                             NaN   

       featureHitRate2AllyHits_value  featureHitRate2HeroHits_value  \
0                                NaN                            NaN   
1                                NaN                            NaN   
2                                NaN                            NaN   
3                                NaN                            NaN   
4                                NaN                            NaN   
...                              ...                            ...   
64051                            NaN                            NaN   
64052                            NaN                            NaN   
64053                            NaN                            NaN   
64054                            NaN                            NaN   
64055                            NaN                            NaN   

       featureHitRate2EnemyHits_value  featureHitRate2ShieldHits_value  \
0                                 NaN                              NaN   
1                                 NaN                              NaN   
2                                 NaN                              NaN   
3                                 NaN                              NaN   
4                                 NaN                              NaN   
...                               ...                              ...   
64051                             NaN                              NaN   
64052                             NaN                              NaN   
64053                       

*****
This row number is fine:  

52,524 rows / 12 players/game = 4,377 games worth of data 
*****

will do the grouping by match id when analyzing data, just how player_id was used to group data in player data notebook

In [7]:
matches_cleaned[matches_cleaned.player_id == "VP Sypeh"]


match_id player_id  team_id  won  \
4      5513730_1765074019_1318548_11001_12  VP Sypeh        0    1   
13     6711722_1764888815_1272257_11001_22  VP Sypeh        0    1   
32     6711647_1764886525_1245250_11001_22  VP Sypeh        1    1   
43      506241_1764884213_1311009_11001_21  VP Sypeh        0    1   
48      506105_1764882596_1310017_11001_21  VP Sypeh        1    0   
71     5513709_1764826259_2042189_11001_12  VP Sypeh        0    1   
74     5513562_1764654765_1291214_11001_12  VP Sypeh        1    1   
88     5513669_1764653691_1231370_11001_12  VP Sypeh        1    1   
105    5513547_1764535377_1311217_11001_12  VP Sypeh        1    1   
117    5513459_1764389811_1291129_11001_12  VP Sypeh        0    1   
124    5513743_1764388496_1310548_11001_12  VP Sypeh        0    1   
137    5513958_1764387072_1290004_11001_12  VP Sypeh        0    1   
150    5513656_1764385747_2042004_11001_12  VP Sypeh        1    1   
163    5513300_1764384135_1231532_11001_12  VP Sypeh        1    0   
170    5513381_1764383145_1310400_11001_12  VP Sypeh        1    1   
180    5513524_1764381382_1267309_11001_12  VP Sypeh        1    0   
192    5513807_1764379510_1288281_11001_12  VP Sypeh        1    0   
207    6711550_1764378652_1272489_11001_22  VP Sypeh        0    0   
223    6711698_1764377320_1231426_11001_22  VP Sypeh        0    1   
229    6711650_1764188729_1267046_11001_22  VP Sypeh        0    1   
242     506090_1763678287_1291100_11001_21  VP Sypeh        1    1   
256    5513668_1763676943_1272043_11001_12  VP Sypeh        0    1   
275    5513375_1763675468_1288104_11001_12  VP Sypeh        0    1   
283    5513662_1763674249_1290011_11001_12  VP Sypeh        1    1   
290    5513612_1763673169_1318044_11001_12  VP Sypeh        0    1   
9463   6711480_1763578022_1310494_11001_22  VP Sypeh        1    1   
9478   6711443_1763335979_1245235_11001_22  VP Sypeh        1    1   
9491   6711433_1763334632_1318088_11001_22  VP Sypeh        1    1   
9498    506127_1763333643_1311233_11001_21  VP Sypeh        1    1   
9508   6711393_1763332165_2042017_11001_22  VP Sypeh        0    1   
17385  6711756_1763249456_1288333_11001_22  VP Sypeh        1    0   
17392  6711506_1763247996_1292231_11001_22  VP Sypeh        1    0   
17405   506016_1763246693_1291465_11001_21  VP Sypeh        1    1   
17420  6711479_1763244590_1231522_11001_22  VP Sypeh        0    1   
17440  6711635_1763241195_1267405_11001_22  VP Sypeh        0    1   
17480   505963_1763237397_1310437_11001_21  VP Sypeh        1    1   
28409  5513804_1763595741_1290073_11001_12  VP Sypeh        0    0   
41308  5513632_1763591081_1311030_11001_12  VP Sypeh        1    0   

             hero_name  hero_id           map_name   match_mode  duration_sec  \
4      Invisible Woman     1050     Celestial Husk   Domination           381   
13     Invisible Woman     1050    Birnin T'Challa   Domination           537   
32     Invisible Woman     1050     Spider-Islands       Convoy          1015   
43              Gambit     1058             Arakko       Convoy          1187   
48     Invisible Woman     1050             Krakoa   Domination           883   
71              Gambit     1058    Heart Of Heaven  Convergence           234   
74     Invisible Woman     1050            Midtown       Convoy           960   
88      Rocket Raccoon     1023    Yggdrasill Path       Convoy           259   
105             Gambit     1058             Arakko       Convoy           749   
117     Rocket Raccoon     1023            Midtown       Convoy           639   
124    Invisible Woman     1050             Krakoa   Domination           515   
137    Invisible Woman     1050  Symbiotic Surface  Convergence           893   
150               Loki     1016    Heart Of Heaven  Convergence           637   
163             Mantis     1020    Yggdrasill Path       Convoy           757   
170    Invisible Woman     1050             Krakoa   Domination           469   
180    

i will select for specific columns, 72 is a bit much, need to see what the fluff is

In [8]:
matches_cleaned.head().to_csv('matches_cleaned_head.csv', index=False)

I looked through the file, found columns that are meaningless for me, and copy pasted below, and put into list to drop

I dont need any of these columns: 
sessionHitRate_value,maxSurvivalKills_value,maxContinueKills_value,maxContinueWins_value,continueKills3_value,continueKills4_value,continueKills5_value,continueKills6_value,mainAttacks_value,mainAttackHits_value,shieldHits_value,summonerHits_value,chaosHits_value,heroSessionHitRate_value,heroMainAttacks_value,heroMainAttackHits_value,heroShieldHits_value,heroSummonerHits_value,heroChaosHits_value,featureNormalData1_value,featureNormalData2_value,featureNormalData3_value,featureCriticalRate1CritHits_value,featureCriticalRate1Hits_value,featureCriticalRate2CritHits_value,featureCriticalRate2Hits_value,featureHitRate1UseCount_value,featureHitRate1RealHitHeroCount_value,featureHitRate1ChaosHits_value,featureHitRate1AllyHits_value,featureHitRate1HeroHits_value,featureHitRate1EnemyHits_value,featureHitRate1ShieldHits_value,featureHitRate1SummonerHits_value,featureHitRate2UseCount_value,featureHitRate2RealHitHeroCount_value,featureHitRate2ChaosHits_value,featureHitRate2AllyHits_value,featureHitRate2HeroHits_value,featureHitRate2EnemyHits_value,featureHitRate2ShieldHits_value,featureHitRate2SummonerHits_value,featureSpecialData1Total_value,featureSpecialData1Value_value,ranked_value,rankedDelta_value

In [9]:
cols_to_drop = [
    "sessionHitRate_value","maxSurvivalKills_value","maxContinueKills_value",
    "maxContinueWins_value","continueKills3_value","continueKills4_value",
    "continueKills5_value","continueKills6_value","mainAttacks_value",
    "mainAttackHits_value","shieldHits_value","summonerHits_value",
    "chaosHits_value","heroSessionHitRate_value","heroMainAttacks_value",
    "heroMainAttackHits_value","heroShieldHits_value","heroSummonerHits_value",
    "heroChaosHits_value","featureNormalData1_value","featureNormalData2_value",
    "featureNormalData3_value","featureCriticalRate1CritHits_value",
    "featureCriticalRate1Hits_value","featureCriticalRate2CritHits_value",
    "featureCriticalRate2Hits_value","featureHitRate1UseCount_value",
    "featureHitRate1RealHitHeroCount_value","featureHitRate1ChaosHits_value",
    "featureHitRate1AllyHits_value","featureHitRate1HeroHits_value",
    "featureHitRate1EnemyHits_value","featureHitRate1ShieldHits_value",
    "featureHitRate1SummonerHits_value","featureHitRate2UseCount_value",
    "featureHitRate2RealHitHeroCount_value","featureHitRate2ChaosHits_value",
    "featureHitRate2AllyHits_value","featureHitRate2HeroHits_value",
    "featureHitRate2EnemyHits_value","featureHitRate2ShieldHits_value",
    "featureHitRate2SummonerHits_value","featureSpecialData1Total_value",
    "featureSpecialData1Value_value","ranked_value","rankedDelta_value"
]

matches_cleaned = matches_cleaned.drop(columns=cols_to_drop)
matches_cleaned

match_id      player_id  team_id  won  \
0      5513730_1765074019_1318548_11001_12     Deleted404        0    1   
1      5513730_1765074019_1318548_11001_12          Kaifs        0    1   
2      5513730_1765074019_1318548_11001_12         Finnsi        0    1   
3      5513730_1765074019_1318548_11001_12    JukeBox2274        1    0   
4      5513730_1765074019_1318548_11001_12       VP Sypeh        0    1   
...                                    ...            ...      ...  ...   
64051  5513925_1765094719_1231224_11001_12          iffvy        1    0   
64052  5513925_1765094719_1231224_11001_12      Rajahyrax        1    0   
64053  5513925_1765094719_1231224_11001_12            Пyк        1    0   
64054  5513925_1765094719_1231224_11001_12  FAZE BLAZIKEN        0    1   
64055  5513925_1765094719_1231224_11001_12  сountry mouse        1    0   

             hero_name  hero_id         map_name  match_mode  duration_sec  \
0       Rocket Raccoon     1023   Celestial Husk  Domination           381   
1              Magneto     1037   Celestial Husk  Domination           381   
2                Groot     1027   Celestial Husk  Domination           381   
3               Angela     1056   Celestial Husk  Domination           381   
4      Invisible Woman     1050   Celestial Husk  Domination           381   
...                ...      ...              ...         ...           ...   
64051   Winter Soldier     1041  Yggdrasill Path      Convoy          1059   
64052            Groot     1027  Yggdrasill Path      Convoy          1059   
64053   Rocket Raccoon     1023  Yggdrasill Path      Convoy          1059   
64054          Phoenix     1054  Yggdrasill Path      Convoy          1059   
64055          Phoenix     1054  Yggdrasill Path      Convoy          1059   

       is_mvp  ...  kdaRatio_value  totalHeroDamage_value  \
0       False  ...       10.000000             655.769723   
1       False  ...       13.000000            7781.313597   
2       False  ...        6.500000            8685.591116   
3       False  ...        0.666667            6113.298075   
4       False  ...       27.000000            4548.798798   
...       ...  ...             ...                    ...   
64051   False  ...        3.818182           32969.463813   
64052   False  ...        4.555556           28770.961880   
64053   False  ...        7.285714            5906.691635   
64054   False  ...        3.200000           38203.973442   
64055   False  ...        2.461538           33735.914978   

       totalHeroDamagePerMinute_value  totalHeroHeal_value  \
0                              109.29         11943.727504   
1                             1296.89             0.000000   
2                             1447.60             0.000000   
3                             1018.88             0.000000   
4                              758.13         14847.611123   
...                               ...                  ...   
64051                         1939.38          2477.660993   
64052                         1692.41             0.000000   
64053                          347.45         39778.870284   
64054                         2247.29          1665.143768   
64055                         1984.47           904.841394   

       totalHeroHealPerMinute_value  totalDamageTaken_value  \
0                           1990.62             2981.606651   
1                              0.00            11832.338579   
2                              0.00            19402.304778   
3                              0.00            18688.655201   
4                           2474.60             6147.800565   
...                             ...                     ...   
64051                        145.74            19305.615006   
64052                          0.00            82469.576583   
64053                       2339.93            12175.875208   
64054                         97.95            15347.804127   
64055                    

Thought process of analyzing team compositions: 
Group by match id, then separate into each team, and which team won. 
Then can do like a large scale analysis of repeated teams and win rates

In [10]:
matches_cleaned.groupby(["hero_name", "map_name"])["won"].mean()


hero_name     map_name         
Adam Warlock  Arakko               0.419355
              Birnin T'Challa      0.473684
              Celestial Husk       0.529412
              Central Park         0.500000
              Hall Of Djalia       0.500000
                                     ...   
Wolverine     Midtown              0.642857
              Royal Palace         0.000000
              Spider-Islands       0.558140
              Symbiotic Surface    0.455882
              Yggdrasill Path      0.459016
Name: won, Length: 539, dtype: float64

In [11]:
matches_cleaned.groupby(["hero_name", "match_mode"])["won"].mean()


hero_name       match_mode 
Adam Warlock    Convergence    0.461538
                Convoy         0.500000
                Domination     0.438596
Angela          Convergence    0.555556
                Convoy         0.550256
                                 ...   
Winter Soldier  Convoy         0.427657
                Domination     0.446271
Wolverine       Convergence    0.535055
                Convoy         0.557927
                Domination     0.542088
Name: won, Length: 132, dtype: float64

In [12]:
matches_cleaned[matches_cleaned.player_id == 'VP Sypeh'].groupby(["hero_name", "match_mode"])["won"].mean()


hero_name        match_mode 
Adam Warlock     Domination     1.000000
Emma Frost       Convoy         1.000000
Gambit           Convergence    0.600000
                 Convoy         0.857143
                 Domination     0.500000
Invisible Woman  Convergence    0.666667
                 Convoy         1.000000
                 Domination     0.714286
Loki             Convergence    1.000000
                 Convoy         1.000000
                 Domination     1.000000
Mantis           Convoy         0.000000
Rocket Raccoon   Convoy         1.000000
Name: won, dtype: float64

the above outcomes match what i see in the per player data, so this is good (to double check the data is valid/reliable in case of discrepancies)

Thought process:  

What i want to analyze with this data set is the team compositions per team (which combination of 6 heroes are there per team) and who won?   

Based on that I want to create like a matchup criteria, where if theres one team composition, how likely are they to win against an enemy team composition.   

And also need to see, if the enemy team composition is unknown, how good is the single team composition's performance in general?

In [13]:
# gonna get specific team comps now into their own rows
team_row_separating = matches_cleaned[['match_id', 'team_id', 'hero_name', 'won']].copy()
team_row_separating

match_id  team_id        hero_name  won
0      5513730_1765074019_1318548_11001_12        0   Rocket Raccoon    1
1      5513730_1765074019_1318548_11001_12        0          Magneto    1
2      5513730_1765074019_1318548_11001_12        0            Groot    1
3      5513730_1765074019_1318548_11001_12        1           Angela    0
4      5513730_1765074019_1318548_11001_12        0  Invisible Woman    1
...                                    ...      ...              ...  ...
64051  5513925_1765094719_1231224_11001_12        1   Winter Soldier    0
64052  5513925_1765094719_1231224_11001_12        1            Groot    0
64053  5513925_1765094719_1231224_11001_12        1   Rocket Raccoon    0
64054  5513925_1765094719_1231224_11001_12        0          Phoenix    1
64055  5513925_1765094719_1231224_11001_12        1          Phoenix    0

[64056 rows x 4 columns]

grouping them up now

In [14]:
#have to sort during the lambda function otherwise the same heroes in different orders will be treated as different comps
#grouping by the match and the 2 teams, and then for each team aggregating the heroes and ordering them w/ comma separation
team_comps = team_row_separating.groupby(['match_id', 'team_id']).agg({'hero_name': lambda x: ', '.join(sorted(x)),
'won': 'max'}) 
team_comps = team_comps.reset_index()
team_comps

match_id  team_id  \
0      4859912_1764413878_2042276_11001_32        0   
1      4859912_1764413878_2042276_11001_32        1   
2      4859918_1764785795_1291050_11001_32        0   
3      4859918_1764785795_1291050_11001_32        1   
4      4859927_1764258769_1291106_11001_32        0   
...                                    ...      ...   
10671  6713333_1763758119_1267033_11001_22        1   
10672  6713340_1764966783_2042019_11001_22        0   
10673  6713340_1764966783_2042019_11001_22        1   
10674      6713853_1736615457_110_11001_10        0   
10675      6713853_1736615457_110_11001_10        1   

                                               hero_name  won  
0      Gambit, Invisible Woman, Magneto, Phoenix, Spi...    1  
1      Gambit, Invisible Woman, Magneto, Spider-Man, ...    0  
2      Daredevil, Gambit, Luna Snow, Magneto, Venom, ...    0  
3      Gambit, Iron Fist, Magneto, Phoenix, Rocket Ra...    1  
4      Blade, Emma Frost, Invisible Woman, Magneto, R...    1  
...                                                  ...  ...  
10671  Black Widow, Gambit, Invisible Woman, Magneto,...    1  
10672  Black Panther, Emma Frost, Invisible Woman, Lu...    0  
10673  Angela, Doctor Strange, Hawkeye, Invisible Wom...    1  
10674  Invisible Woman, Iron Man, Mantis, Mister Fant...    1  
10675  Bruce Banner, Cloak & Dagger, Hawkeye, Invisib...    0  

[10676 rows x 4 columns]

each match id is listed 2 times (one row for each team comp)

In [15]:
team_comps.to_csv('team_comps.csv', index=False)

now that the matches have the set of heroes on each team, I want to see what the win rate in general of each team composition, with each team composition being the set of heroes under the 'hero_name' column

In [16]:
win_rates = team_comps.groupby('hero_name')['won'].mean().reset_index()
win_rates = win_rates.rename(columns={'won': 'win_rate'})
win_rates

hero_name  win_rate
0     Adam Warlock, Adam Warlock, Bruce Banner, Gamb...       0.0
1     Adam Warlock, Angela, Bruce Banner, Gambit, In...       1.0
2     Adam Warlock, Angela, Captain America, Cloak &...       0.0
3     Adam Warlock, Angela, Captain America, Daredev...       0.0
4     Adam Warlock, Angela, Captain America, Hela, J...       0.0
...                                                 ...       ...
7834  Magneto, Moon Knight, Psylocke, Rocket Raccoon...       1.0
7835  Magneto, Peni Parker, Phoenix, Psylocke, Rocke...       0.0
7836  Magneto, Phoenix, Rocket Raccoon, Spider-Man, ...       0.0
7837  Mantis, Psylocke, Rocket Raccoon, Thor, Venom,...       1.0
7838  Peni Parker, Phoenix, Psylocke, Rocket Raccoon...       1.0

[7839 rows x 2 columns]

the above doesnt look right, is the list of hero names is being treated like an ordered list instead of a set?  

How come theres 7839 unique team compositions? I find that wierd

In [17]:
win_rates.to_csv('team_comp_win_rates.csv', index=False)

Actually, looking through the csv file, it actually does look right. It is very crazy yto see that there can be so many unique team compositions, but it is unsurprising when considering the total #heroes in the game

The math does check out though, as there are currently 44 heroes in the game, so the total combinations of heroes possible per team is a pretty high number. 

I want the total count a team composition appears in the table so i know whether 100% winrate is from 30 matches or 1

In [18]:
counts = team_comps['hero_name'].value_counts().reset_index()
counts.columns = ['hero_name', 'game_count']
win_rates = win_rates.merge(counts, on='hero_name')
win_rates

hero_name  win_rate  game_count
0     Adam Warlock, Adam Warlock, Bruce Banner, Gamb...       0.0           1
1     Adam Warlock, Angela, Bruce Banner, Gambit, In...       1.0           1
2     Adam Warlock, Angela, Captain America, Cloak &...       0.0           1
3     Adam Warlock, Angela, Captain America, Daredev...       0.0           1
4     Adam Warlock, Angela, Captain America, Hela, J...       0.0           1
...                                                 ...       ...         ...
7834  Magneto, Moon Knight, Psylocke, Rocket Raccoon...       1.0           1
7835  Magneto, Peni Parker, Phoenix, Psylocke, Rocke...       0.0           1
7836  Magneto, Phoenix, Rocket Raccoon, Spider-Man, ...       0.0           1
7837  Mantis, Psylocke, Rocket Raccoon, Thor, Venom,...       1.0           1
7838  Peni Parker, Phoenix, Psylocke, Rocket Raccoon...       1.0           1

[7839 rows x 3 columns]

In [19]:
matches_cleaned.to_csv('matches_cleaned.csv', index=False)
win_rates.to_csv('team_comp_win_rates.csv', index=False)